### PR TITLE
Implement `component.cached`

### PR DESCRIPTION
**Pull request by Gozala**
Implementation for #64

### CODE REVIEW

**Comment by Gozala** [comment]
Still need to write tests to actually test it does the job, but wanted to share regardless to gather a feedback.

**Comment by mikaelbr** [comment]
Looks good. But I don't think `isEqualProps` is the currect method to use here. As it would fail if you pass in a single cursor.

File: component.js [hunk 1]
```diff
@@ -102,6 +103,7 @@ function factory (options) {
   var _isJsx = !!options.jsx;
   var _hiddenCursorField = options.cursorField || '__singleCursor';
   var _isNode = options.isNode || isNode;
+  var _cached = cached.withDefaults(_shouldComponentUpdate);
```
Comment (mikaelbr) [inline component.js]: This isn't right? This should probably just be `cached.withDefaults(options)`, given that the function used (currently `isEqualProps`) is descriptive for this use case and it's clear that changing default on Omniscient also changes default on cached. `isEqualProps` won't work for the reasons described in a previous comment.

File: cached.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+   * not practical to call it mulitple times with in the same `render`
+   * call.
+   *
+   * @param {Function} function doing computation
+   *
+   * @module shouldComponentUpdate.isEqualState
+   * @returns {Function} Optimized function.
+   * @api public
+   */
+  return function cached (f) {
```
Comment (mikaelbr) [inline cached.js]: This will imply that you need to declare the cached function outside render, and not create a new cache every render right?

**Comment by Gozala** [comment]
> Looks good. But I don't think isEqualProps is the currect method to use here. As it would fail if you pass in a single cursor.

I don't think so as we compare argument objects not cursors themselves so it should be fine. Either way I'll have test case to cover this ;)

File: cached.js (hunk 2 above)
Comment (Gozala) [inline cached.js]: That is correct.

File: component.js (hunk 1 above)
Comment (Gozala) [inline component.js]: Just to be clear I understand, what you suggesting to do is:

```js
var _cached = cached.withDefaults(options);
```

And then in `cached` factory do:

```js
function factory (methods) {
  var isEqualProps = methods. isEqualProps || shouldUpdate.isEqualProps;
}
```

Is my understanding correct ?

Also I believe `isEqualProps` will work for the reason I've posted earlier.

**Comment by mikaelbr** [comment]
Very true. Didn't catch that at first glance. :+1:

File: component.js (hunk 1 above)
Comment (mikaelbr) [inline component.js]: Now you do 

```js
function factory (methods) {
  var isEqual = (methods && methods.isEqualProps) || shouldupdate.isEqualProps;
  // ...
```

And passing in the function like:

```js
var _cached = cached.withDefaults(_shouldComponentUpdate);
```

Wouldn't work. You'd always use `shouldupdate.isEqualProps`. But passing in `options` instead, would work.

yeah, `isEqualProps` would work! :)

**Comment by Gozala** [comment]
@mikaelbr @torgeir I have pushed tests for this new functionality. Please take a look and let me know if it looks ready to be merged or if there is anything else to be done to get there.

**Comment by torgeir** [comment]
nice!

**Pull request merged**